### PR TITLE
fix(history): fix duplicate documents

### LINKED ISSUE #798: Duplicates in historical responses
When querying the historical API I get a duplicate entry in the `fromRevision`.

Apparently, the entry only exists once in the database.

```python
>>> d = Document.objects.get(pk="44fcbaf9-1892-4945-a48f-cd4cba8ee138")
>>> d.history.filter(history_date="2019-11-14 16:54:39.874867+00:00")
<QuerySet [<HistoricalDocument: Document object (44fcbaf9-1892-4945-a48f-cd4cba8ee138) as of 2019-11-14 16:54:39.874867+00:00>]>
```

```json
{
  "id": 
"SGlzdG9yaWNhbERvY3VtZW50OmQ1MTk0MmZhLWFiZGQtNGM3OC1hZjU0LThiMmNiOWQ2NmZlYQ==",
  "documentId": "44fcbaf9-1892-4945-a48f-cd4cba8ee138",
  "historyDate": "2019-11-14T16:54:39.874867+00:00",
  "historyType": "~",
  "historyUserId": "855a5ba0-cdcc-48d5-ab7c-4fc55cbe0754",
  "__typename": "HistoricalDocument",
  "historicalAnswers": {
    "edges": [
      {
        "node": {
          "historyDate": "2019-11-14T16:54:03.009650+00:00",
          "historyUserId": "855a5ba0-cdcc-48d5-ab7c-4fc55cbe0754",
          "historyType": "+",
          "question": {
            "slug": "name-gesuchstellerin",
            "label": "Name",
            "__typename": "TextQuestion"
          },
          "stringValue": "A",
          "__typename": "HistoricalStringAnswer"
        }
      },
      {
        "node": {
          "historyDate": "2019-11-14T16:54:07.083435+00:00",
          "historyUserId": "855a5ba0-cdcc-48d5-ab7c-4fc55cbe0754",
          "historyType": "+",
          "question": {
            "slug": "plz-gesuchstellerin",
            "label": "PLZ",
            "__typename": "IntegerQuestion"
          },
          "integerValue": 1111,
          "__typename": "HistoricalIntegerAnswer"
        }
      },
      {
        "node": {
          "historyDate": "2019-11-14T16:54:03.637712+00:00",
          "historyUserId": "855a5ba0-cdcc-48d5-ab7c-4fc55cbe0754",
          "historyType": "+",
          "question": {
            "slug": "vorname-gesuchstellerin",
            "label": "Vorname",
            "__typename": "TextQuestion"
          },
          "stringValue": "A",
          "__typename": "HistoricalStringAnswer"
        }
      },
      {
        "node": {
          "historyDate": "2019-11-14T16:54:04.319241+00:00",
          "historyUserId": "855a5ba0-cdcc-48d5-ab7c-4fc55cbe0754",
          "historyType": "+",
          "question": {
            "slug": "strasse-gesuchstellerin",
            "label": "Strasse",
            "__typename": "TextQuestion"
          },
          "stringValue": "A",
          "__typename": "HistoricalStringAnswer"
        }
      },
      {
        "node": {
          "historyDate": "2019-11-14T16:54:04.995446+00:00",
          "historyUserId": "855a5ba0-cdcc-48d5-ab7c-4fc55cbe0754",
          "historyType": "+",
          "question": {
            "slug": "nummer-gesuchstellerin",
            "label": "Nummer",
            "__typename": "TextQuestion"
          },
          "stringValue": "A",
          "__typename": "HistoricalStringAnswer"
        }
      },
      {
        "node": {
          "historyDate": "2019-11-14T16:54:10.580440+00:00",
          "historyUserId": "855a5ba0-cdcc-48d5-ab7c-4fc55cbe0754",
          "historyType": "+",
          "question": {
            "slug": "telefon-oder-mobile-gesuchstellerin",
            "label": "Telefon oder Mobile",
            "__typename": "TextQuestion"
          },
          "stringValue": "0111111",
          "__typename": "HistoricalStringAnswer"
        }
      },
      {
        "node": {
          "historyDate": "2019-11-14T16:54:07.850998+00:00",
          "historyUserId": "855a5ba0-cdcc-48d5-ab7c-4fc55cbe0754",
          "historyType": "+",
          "question": {
            "slug": "ort-gesuchstellerin",
            "label": "Ort",
            "__typename": "TextQuestion"
          },
          "stringValue": "A",
          "__typename": "HistoricalStringAnswer"
        }
      },
      {
        "node": {
          "historyDate": "2019-11-14T16:54:17.492774+00:00",
          "historyUserId": "855a5ba0-cdcc-48d5-ab7c-4fc55cbe0754",
          "historyType": "~",
          "question": {
            "slug": "e-mail-gesuchstellerin",
            "label": "E-Mail",
            "__typename": "TextQuestion"
          },
          "stringValue": "A@A.A",
          "__typename": "HistoricalStringAnswer"
        }
      },
      {
        "node": {
          "historyDate": "2019-11-14T16:54:00.837669+00:00",
          "historyUserId": "855a5ba0-cdcc-48d5-ab7c-4fc55cbe0754",
          "historyType": "+",
          "question": {
            "slug": "juristische-person-gesuchstellerin",
            "label": "Handelt es sich um  eine juristische Person?",
            "choiceOptions": {
              "edges": [
                {
                  "node": {
                    "slug": "juristische-person-gesuchstellerin-ja",
                    "label": "Ja"
                  }
                },
                {
                  "node": {
                    "slug": "juristische-person-gesuchstellerin-nein",
                    "label": "Nein"
                  }
                }
              ]
            },
            "__typename": "ChoiceQuestion"
          },
          "stringValue": "juristische-person-gesuchstellerin-nein",
          "__typename": "HistoricalStringAnswer"
        }
      }
    ]
  }
},
{
  "id": 
"SGlzdG9yaWNhbERvY3VtZW50OmQ1MTk0MmZhLWFiZGQtNGM3OC1hZjU0LThiMmNiOWQ2NmZlYQ==",
  "documentId": "44fcbaf9-1892-4945-a48f-cd4cba8ee138",
  "historyDate": "2019-11-14T16:54:39.874867+00:00",
  "historyType": "~",
  "historyUserId": "855a5ba0-cdcc-48d5-ab7c-4fc55cbe0754",
  "__typename": "HistoricalDocument",
  "historicalAnswers": {
    "edges": [
      {
        "node": {
          "historyDate": "2019-11-14T16:54:03.009650+00:00",
          "historyUserId": "855a5ba0-cdcc-48d5-ab7c-4fc55cbe0754",
          "historyType": "+",
          "question": {
            "slug": "name-gesuchstellerin",
            "label": "Name",
            "__typename": "TextQuestion"
          },
          "stringValue": "A",
          "__typename": "HistoricalStringAnswer"
        }
      },
      {
        "node": {
          "historyDate": "2019-11-14T16:54:07.083435+00:00",
          "historyUserId": "855a5ba0-cdcc-48d5-ab7c-4fc55cbe0754",
          "historyType": "+",
          "question": {
            "slug": "plz-gesuchstellerin",
            "label": "PLZ",
            "__typename": "IntegerQuestion"
          },
          "integerValue": 1111,
          "__typename": "HistoricalIntegerAnswer"
        }
      },
      {
        "node": {
          "historyDate": "2019-11-14T16:54:03.637712+00:00",
          "historyUserId": "855a5ba0-cdcc-48d5-ab7c-4fc55cbe0754",
          "historyType": "+",
          "question": {
            "slug": "vorname-gesuchstellerin",
            "label": "Vorname",
            "__typename": "TextQuestion"
          },
          "stringValue": "A",
          "__typename": "HistoricalStringAnswer"
        }
      },
      {
        "node": {
          "historyDate": "2019-11-14T16:54:04.319241+00:00",
          "historyUserId": "855a5ba0-cdcc-48d5-ab7c-4fc55cbe0754",
          "historyType": "+",
          "question": {
            "slug": "strasse-gesuchstellerin",
            "label": "Strasse",
            "__typename": "TextQuestion"
          },
          "stringValue": "A",
          "__typename": "HistoricalStringAnswer"
        }
      },
      {
        "node": {
          "historyDate": "2019-11-14T16:54:04.995446+00:00",
          "historyUserId": "855a5ba0-cdcc-48d5-ab7c-4fc55cbe0754",
          "historyType": "+",
          "question": {
            "slug": "nummer-gesuchstellerin",
            "label": "Nummer",
            "__typename": "TextQuestion"
          },
          "stringValue": "A",
          "__typename": "HistoricalStringAnswer"
        }
      },
      {
        "node": {
          "historyDate": "2019-11-14T16:54:10.580440+00:00",
          "historyUserId": "855a5ba0-cdcc-48d5-ab7c-4fc55cbe0754",
          "historyType": "+",
          "question": {
            "slug": "telefon-oder-mobile-gesuchstellerin",
            "label": "Telefon oder Mobile",
            "__typename": "TextQuestion"
          },
          "stringValue": "0111111",
          "__typename": "HistoricalStringAnswer"
        }
      },
      {
        "node": {
          "historyDate": "2019-11-14T16:54:07.850998+00:00",
          "historyUserId": "855a5ba0-cdcc-48d5-ab7c-4fc55cbe0754",
          "historyType": "+",
          "question": {
            "slug": "ort-gesuchstellerin",
            "label": "Ort",
            "__typename": "TextQuestion"
          },
          "stringValue": "A",
          "__typename": "HistoricalStringAnswer"
        }
      },
      {
        "node": {
          "historyDate": "2019-11-14T16:54:17.492774+00:00",
          "historyUserId": "855a5ba0-cdcc-48d5-ab7c-4fc55cbe0754",
          "historyType": "~",
          "question": {
            "slug": "e-mail-gesuchstellerin",
            "label": "E-Mail",
            "__typename": "TextQuestion"
          },
          "stringValue": "A@A.A",
          "__typename": "HistoricalStringAnswer"
        }
      },
      {
        "node": {
          "historyDate": "2019-11-14T16:54:00.837669+00:00",
          "historyUserId": "855a5ba0-cdcc-48d5-ab7c-4fc55cbe0754",
          "historyType": "+",
          "question": {
            "slug": "juristische-person-gesuchstellerin",
            "label": "Handelt es sich um  eine juristische Person?",
            "choiceOptions": {
              "edges": [
                {
                  "node": {
                    "slug": "juristische-person-gesuchstellerin-ja",
                    "label": "Ja"
                  }
                },
                {
                  "node": {
                    "slug": "juristische-person-gesuchstellerin-nein",
                    "label": "Nein"
                  }
                }
              ]
            },
            "__typename": "ChoiceQuestion"
          },
          "stringValue": "juristische-person-gesuchstellerin-nein",
          "__typename": "HistoricalStringAnswer"
        }
      }
    ]
  }
}
```

### PR DESCRIPTION
**Pull request by open-dynaMIX**
This commit prevents duplicate documents in `HistoricalTableAnswer`.

Closes #798

### CODE REVIEW

File: caluma/form/historical_schema.py [hunk 1]
```diff
@@ -190,13 +190,13 @@ def resolve_value(self, info, as_of, *args):
...
-            )[0]
-            for ad in answerdocuments
+            models.Document.history.filter(
+                id=ad.document_id, history_date__lte=as_of
+            ).latest("history_date")
+            for ad in set(answerdocuments)
         ]
 
-        return documents
+        return set(documents)
```
Comment (winged) [inline caluma/form/historical_schema.py]: You will lose sorting if you put it in a set, as a set is an unordered data structure.

If you want to preserve ordering, you'd need something like `return sorted(set(documents), key=...)`
instead

Comment (open-dynaMIX) [reply caluma/form/historical_schema.py]: Great find! It's fixed.

File: caluma/form/historical_schema.py [hunk 2]
```diff
@@ -190,13 +190,13 @@ def resolve_value(self, info, as_of, *args):
...
 
         documents = [
-            models.Document.history.filter(id=ad.document_id).filter(
-                history_date__lte=as_of
-            )[0]
-            for ad in answerdocuments
+            models.Document.history.filter(
+                id=ad.document_id, history_date__lte=as_of
+            ).latest("history_date")
+            for ad in list(dict.fromkeys(answerdocuments))
```
Comment (winged) [inline caluma/form/historical_schema.py]: `AnswerDocument`s and `Document`s can be dict keys? Didn't know!

I'd like a comment to explain how/why this works though. Might confuse future readers

Comment (open-dynaMIX) [reply caluma/form/historical_schema.py]: Done.

Comment (winged) [reply caluma/form/historical_schema.py]: Awesome! Just for future reference: https://docs.python.org/3.6/whatsnew/3.6.html#whatsnew36-compactdict

**Pull request merged**